Only update the package cache on fresh installations

yubiuser opened the pull request:
**What does this PR aim to accomplish?:**

Only update the package cache once on fresh installs, not on checkouts/updates. Addresses https://github.com/pi-hole/pi-hole/discussions/6272

---
**By submitting this pull request, I confirm the following:**

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against.
2. I have commented my proposed changes within the code and I have tested my changes.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))
6. I have checked that another pull request for this purpose does not exist.
7. I have considered, and confirmed that this submission will be valuable to others.
8. I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
9. I give this submission freely, and claim no ownership to its content.

---
- [x] I have read the above and my PR is ready for review. *Check this box to confirm*

yubiuser commented:
The seconds commit should fix the tests. Not sure why we had `fresh_install=false` in a test that should test a fresh installation.

darkexplosiveqwx commented:
I think we should still update the package cache when the meta package is updated, so all new dependencies are up to date.

yubiuser commented:
In the current implementation or the one suggestion by @darkexplosiveqwx ?

https://github.com/pi-hole/pi-hole/pull/6274 removes also the building and installing of the dependency package before doing an update - which might fails because some 'installer dependency' is not installed.

mwoolweaver commented:
> Only update the package cache once on fresh installs, not on checkouts/updates.

Package cache should also be updated when `pihole-meta` package is being upgraded, right?

mwoolweaver commented:
could you use the following to get version number of new meta package after being it's created. This command would keep from need another variable for the meta package version number. 

```
dpkg-deb -f pihole-meta.deb Version
```

and then compare that to the the installed version found with

```
dpkg -s pihole-meta | grep '^Version:' | cut -d' ' -f2
```

if they are equal then the package cache update can be skipped as there is no way the meta package would be installed if any of its depends aren't installed.

yubiuser commented:
It was added in the first place with this commit
https://github.com/pi-hole/pi-hole/commit/cedd1a259105dc231744d58562f0bc80d4294c7a

Back then, the test added configuration to setupVars.conf to automatically install the web interace and the web server. This is no longer needed.

https://github.com/pi-hole/pi-hole/blob/cedd1a259105dc231744d58562f0bc80d4294c7a/test/test_automated_install.py#L198-L214

mwoolweaver commented:
Would #6292 be a more graceful solution?

yubiuser commented:
I'm not sure if this is the way to go. While it addresses the suggestions from above, it adds 100 lines of code. I'm not sure if this is worth it, considering 'normal' users will only benefit from it during an update which is not happening that often.

___

I would go another way as a middle ground: only update the cache before we install the meta package initiated by the install script, but not from the updater.

yubiuser commented:
The last two changes allow to update the cache when building/installing the meta package from the installer script, but don't do when building/installing from the update script. 
I intentionally did not squash to allow easy reverting to a previous state depending on how we decide the functions should  work.

mwoolweaver commented:
> While it addresses the suggestions from above, it adds 100 lines of code.

But it saves about 160 lines with each `pihole -up` that contains no meta data changes and saves writes to SD card caused by installing the same package again. (No updates available) 

If there is an update to pi-hole core repo with NO meta data or just an FTL update it would save double the lines and double the SD activity.

mwoolweaver commented:
#6292 also has the added benefit of negating the issue mentioned [here](https://github.com/pi-hole/pi-hole/pull/6262#issuecomment-2925435842)

dschaper commented:
I'd like to go for this PR for now as it improves the situation. The massive change in #6298 needs some very thorough review and testing and that's going to take time. 

Considering this is a QOL fix and not a bug fix, I'm good with what Yubi has written.

dschaper commented:
@mwoolweaver:

> But it saves about 160 lines with each `pihole -up` that contains no meta data changes

Where do you see these 160 lines?

This is the output that I see with this branch code:

```
root@pihole:/etc/.pihole# pihole -up
  [✓] Building dependency package pihole-meta.deb
  [✓] Installing Pi-hole dependency package

  [i] Checking for updates...
  [i] Pi-hole Core:     up to date
  [i] Web Interface:    up to date
  [i] FTL:              up to date

  [✓] Everything is up to date!
```

I think I like the idea of always creating/updating the dependency package. It gives us a guarantee that the update will function because it will add the packages that we know we need to run the update. This accounts for the case where a user removes a package that we need from the system, and it's really a light process to run. I don't see it generating a lot of writes to any drive or storage media.



Here's a verbose run of the update code:
```bash
+ build_dependency_package
+ rm -rf '/tmp/pihole-meta_*'
+ local tempdir
++ mktemp --directory /tmp/pihole-meta_XXXXX
+ tempdir=/tmp/pihole-meta_JaZox
+ chmod 0755 /tmp/pihole-meta_JaZox
+ is_command apt-get
+ local check_command=apt-get
+ command -v apt-get
+ pushd /tmp
+ rm -f /tmp/pihole-meta.deb
+ mkdir -p /tmp/pihole-meta_JaZox/DEBIAN
+ chmod 0755 /tmp/pihole-meta_JaZox/DEBIAN
+ touch /tmp/pihole-meta_JaZox/DEBIAN/control
+ echo 'Package: pihole-meta
Version: 0.4
Maintainer: Pi-hole team <adblock@pi-hole.net>
Architecture: all
Description: Pi-hole dependency meta package
Depends: awk,bash-completion,binutils,ca-certificates,cron|cron-daemon,curl,dialog,dnsutils,dns-root-data,git,grep,iproute2,iputils-ping,jq,libcap2,libcap2-bin,lshw,netcat-openbsd,procps,psmisc,sudo,unzip
Section: contrib/metapackages
Priority: optional'
+ local 'str=Building dependency package pihole-meta.deb'
+ printf '  %b %s...' '[i]' 'Building dependency package pihole-meta.deb'
  [i] Building dependency package pihole-meta.deb...+ dpkg-deb --build --root-owner-group /tmp/pihole-meta_JaZox pihole-meta.deb
+ printf '%b  %b %s\n' '\r' '[✓]' 'Building dependency package pihole-meta.deb'
  [✓] Building dependency package pihole-meta.deb
+ popd
+ rm -rf /tmp/pihole-meta_JaZox
+ install_dependent_packages
+ local 'str=Installing Pi-hole dependency package'
+ printf '  %b %s...' '[i]' 'Installing Pi-hole dependency package'
  [i] Installing Pi-hole dependency package...+ is_command apt-get
+ local check_command=apt-get
+ command -v apt-get
+ '[' -f /tmp/pihole-meta.deb ']'
+ eval 'apt-get -qq --no-install-recommends install' /tmp/pihole-meta.deb
+ printf '%b  %b %s\n' '\r' '[✓]' 'Installing Pi-hole dependency package'
  [✓] Installing Pi-hole dependency package
+ rm /tmp/pihole-meta.deb
+ printf '\n'
```

mwoolweaver commented:
@dschaper 

> Where do you see these 160 lines?

the update script calls these 3 functions when installing the meta package

https://github.com/pi-hole/pi-hole/blob/cef7fd4b02c9ae4314e7478c0d542379bfce7f65/advanced/Scripts/update.sh#L116-L119

`package_manager_detect()` is ~38 lines

`build_dependency_package()` is ~92 lines

`install_dependent_packages()` is ~46 lines

so total skipped is 176 lines.

This also doesn't count the calls to other programs used to build the package and install it.

see the `time` test I posted [here](https://github.com/pi-hole/pi-hole/pull/6298#issuecomment-2959931810) for example of time (and likely resources) saved.

Please try #6383 (_this is basically #6298 after following suggestions made by @yubiuser_)

-----

> I think I like the idea of always creating/updating the dependency package.

in an earlier comment you mentioned the following

> >  Maybe something like apt cache policy <dependency package> | grep version and compare that to the /etc/.pihole/ <package metadata version> and "do things" if the proposed updated metadata is higher than the installed package version?
>  >
>  > _Originally posted by dschaper in https://github.com/pi-hole/pi-hole/issues/6282#issuecomment-2944867806_

#6383 accomplishes this
